dot/network/notifications.go

```diff
@@ -349,18 +312,103 @@ func (s *Service) sendData(peer peer.ID, hs Handshake, info *notificationsProtoc
 
 	// we've completed the handshake with the peer, send message directly
 	logger.Tracef("sending message to peer %s using protocol %s: %s", peer, info.protocolID, msg)
-
-	err := s.host.writeToStream(hsData.stream, msg)
-	if err != nil {
+	if err := s.host.writeToStream(stream, msg); err != nil {
 		logger.Debugf("failed to send message to peer %s: %s", peer, err)
+
+		// the stream was closed or reset, close it on our end and delete it from our peer's data
+		if errors.Is(err, io.EOF) || errors.Is(err, mux.ErrReset) {
+			closeOutboundStream(info, peer, stream)
+		}
+		return
 	}
 
+	logger.Tracef("successfully sent message on protocol %s to peer %s: message=", info.protocolID, peer, msg)
 	s.host.cm.peerSetHandler.ReportPeer(peerset.ReputationChange{
 		Value:  peerset.GossipSuccessValue,
 		Reason: peerset.GossipSuccessReason,
 	}, peer)
 }
 
+func (s *Service) sendHandshake(peer peer.ID, hs Handshake, info *notificationsProtocol) (libp2pnetwork.Stream, error) {
+	mu, has := info.outboundHandshakeMutexes.Load(peer)
+	if !has || mu == nil {
```

Comment:
Do we need both checks? Doesn't `!has` suffice?

Reply:
yeah, updated